src/transformers/models/gpt2/modeling_gpt2.py

```diff
@@ -176,11 +190,66 @@ def prune_heads(self, heads):
         self.pruned_heads = self.pruned_heads.union(heads)
 
     def _attn(self, query, key, value, attention_mask=None, head_mask=None):
+
         attn_weights = torch.matmul(query, key.transpose(-1, -2))
 
         if self.scale_attn_weights:
             attn_weights = attn_weights / (float(value.size(-1)) ** 0.5)
 
+        # Layer-wise attention scaling
+        if self.scale_attn_by_layer_idx:
+            attn_weights = attn_weights / float(self.layer_idx + 1)
+
+        if not self.is_cross_attention:
+            # if only "normal" attention layer implements causal mask
+            query_length, key_length = query.size(-2), key.size(-2)
+            causal_mask = self.bias[:, :, key_length - query_length : key_length, :key_length].bool()
+            attn_weights = torch.where(causal_mask, attn_weights, self.masked_bias.to(attn_weights.dtype))
+
+        if attention_mask is not None:
+            # Apply the attention mask
+            attn_weights = attn_weights + attention_mask
+
+        attn_weights = nn.Softmax(dim=-1)(attn_weights)
+
+        # Downcast (if necessary) back to V dtype (half/fp16 if mixed-precision) -- No-Op if in float()
```

Comment:
```suggestion
        # Downcast (if necessary) back to V's dtype if in mixed-precision, No-Op otherwise
```

half == fp16, no need to repeat; but also bf16 is emerging so let's not hardcode fp16. as mixed precision can be either fp16 or bf16.